File: leetcode/tree/865. Smallest Subtree with all the Deepest Nodes/js/main.js
/*

Given a binary tree rooted at root, the depth of each node is the shortest distance to the root.

A node is deepest if it has the largest depth possible among any node in the entire tree.

The subtree of a node is that node, plus the set of all descendants of that node.

Return the node with the largest depth such that it contains all the deepest nodes in its subtree.

Example 1:
  Input: [3,5,1,6,2,0,8,null,null,7,4]
  Output: [2,7,4]

Explanation:
  We return the node with value 2, colored in yellow in the diagram.
  The nodes colored in blue are the deepest nodes of the tree.
  The input "[3, 5, 1, 6, 2, 0, 8, null, null, 7, 4]" is a serialization of the given tree.
  The output "[2, 7, 4]" is a serialization of the subtree rooted at the node with value 2.
  Both the input and output have TreeNode type.
  
Note:
  The number of nodes in the tree will be between 1 and 500.
  The values of each node are unique.

*/

// Time O(N)
// Space O(N)
const subtreeWithAllDeepest = root => {
  if (root == null || isLeaf(root)) return root;

  let ans = null;
  let maxDepth = getMaxDepth(root);

  dfs(root, 1);

  return ans;

  function dfs(node, d) {
    if (node == null) {
      return null;
    }

    if (d == maxDepth) {
      ans = node;
      return node;
    }

    if (isLeaf(node)) {
      return null;
    }

    let l = dfs(node.left, d + 1);
    let r = dfs(node.right, d + 1);

    if (l != null && r != null) {
      ans = node;
    }

    return l || r;
  }

  function isLeaf(node) {
    return node.left == null && node.right == null;
  }

  function getMaxDepth(node) {
    if (node == null) return 0;
    let l = getMaxDepth(node.left);
    let r = getMaxDepth(node.right);
    return Math.max(l, r) + 1;
  }
};

File: leetcode/tree/865. Smallest Subtree with all the Deepest Nodes/py/main.py
import collections


class Solution:
    def __init__(self):
        self.ans = None

    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        if root == None or self.isLeaf(root):
            return root

        maxDepth = self.getMaxDepth(root)

        self.dfs(root, 1, maxDepth)

        return self.ans

    def dfs(self, root, depth, maxDepth):
        if root == None:
            return None

        if depth == maxDepth:
            self.ans = root
            return root

        L = self.dfs(root.left, depth + 1, maxDepth)
        R = self.dfs(root.right, depth + 1, maxDepth)

        if L and R:
            self.ans = root

        return L or R

    def getMaxDepth(self, root):
        if root == None:
            return 0

        l = self.getMaxDepth(root.left)
        r = self.getMaxDepth(root.right)

        return max(l, r) + 1

    def isLeaf(self, root):
        return root.left == None and root.right == None


class Solution_II:
    # Time O(N)
    # Space O(N)
    def subtreeWithAllDeepest(self, root: TreeNode) -> TreeNode:
        depths = collections.defaultdict(list)
        maxDepth = 0
        parent = {}

        def dfs(node, p, d):
            nonlocal maxDepth

            if node is None:
                return

            depths[d].append(node)
            maxDepth = max(maxDepth, d)

            parent[node] = p

            dfs(node.left, node, d + 1)
            dfs(node.right, node, d + 1)

        dfs(root, None, 0)

        queue = collections.deque()
        visited = {}

        for n in depths[maxDepth]:
            queue.append(n)

        target = len(queue)

        while queue:
            size = len(queue)

            for _ in range(size):
                n = queue.popleft()

                if n in parent and parent[n] != None:
                    queue.append(parent[n])
                    visited[n] = visited.get(n, 0) + 1

                    if visited[n] == target:
                        return n

        return root
